In [2]:
import pandas as pd
import numpy as np 
import os
os.chdir("/Users/zed/VSCode/Stats-Alg/HW")

In [3]:
data = pd.read_table("CompanyBill.txt",header=None)
data[data == 0] =np.nan
data = data.dropna(axis=0, how='any')

## 1

In [4]:
print("rows:",data.shape[0],"columns:",data.shape[1])

rows: 4266 columns: 7


## 2

In [5]:
cindex = []
for i in range(1,8):
    cindex.append("V"+str(i))
data.columns = cindex
data.to_csv("data.csv",index=False)

In [6]:
from scipy import stats
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

In [7]:
model = ols("V1~V2+V3+V4+V5+V6+V7",data = data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     V1   R-squared:                       0.989
Model:                            OLS   Adj. R-squared:                  0.989
Method:                 Least Squares   F-statistic:                 6.421e+04
Date:                Sun, 06 Dec 2020   Prob (F-statistic):               0.00
Time:                        20:01:44   Log-Likelihood:                -52328.
No. Observations:                4266   AIC:                         1.047e+05
Df Residuals:                    4259   BIC:                         1.047e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -9069.3257   2297.411     -3.948      0.000   -1.36e+04   -4565.203
V2             0.3902      0.009     44.526      0.000       0.373       0.407
V3             0.3186      0.011     30.133      0.000       0.298       0.339
V4            -0.0947      0.018     -5.319      0.000      -0.130      -0.060
V5             0.5268      0.025     20.734      0.000       0.477       0.577
V6            -0.0883      0.023     -3.844      0.000      -0.133      -0.043
V7           484.7038     87.295      5.552      0.000     313.559     655.848
==============================================================================
Omnibus:                     3740.589   Durbin-Watson:                   1.835
Prob(Omnibus):                  0.000   Jarque-Bera (JB):         13015756.569
Skew:                           2.704   Prob(JB):                         0.00
Kurtosis:                     273.547   Cond. No.                     3.30e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.3e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

As the Adj-$R^2$ is 0.989，it's too high for a real model.AIC and BIC is also too high,there is large probability of the model to overfit the data.

In [9]:
model1=ols(formula='V1~V2+V3+V4',data=data).fit()
model2=ols(formula='V1~V2+V3+V5',data=data).fit()
model3=ols(formula='V1~V2+V4+V5',data=data).fit()
model4=ols(formula='V1~V3+V4+V5',data=data).fit()
print(model1.summary())
print(model2.summary())
print(model3.summary())
print(model4.summary())

                            OLS Regression Results                            
Dep. Variable:                     V1   R-squared:                       0.987
Model:                            OLS   Adj. R-squared:                  0.987
Method:                 Least Squares   F-statistic:                 1.122e+05
Date:                Sun, 06 Dec 2020   Prob (F-statistic):               0.00
Time:                        20:02:03   Log-Likelihood:                -52613.
No. Observations:                4266   AIC:                         1.052e+05
Df Residuals:                    4262   BIC:                         1.053e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4130.6136    987.569      4.183      0.0

The new model is better,since the AIC and BIC is low than former.

In [11]:
model5=ols(formula='V1~V2+V3+V5+V2*V3+V2*V5+V3*V5',data=data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     V1   R-squared:                       0.989
Model:                            OLS   Adj. R-squared:                  0.989
Method:                 Least Squares   F-statistic:                 6.421e+04
Date:                Sun, 06 Dec 2020   Prob (F-statistic):               0.00
Time:                        20:02:51   Log-Likelihood:                -52328.
No. Observations:                4266   AIC:                         1.047e+05
Df Residuals:                    4259   BIC:                         1.047e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -9069.3257   2297.411     -3.948      0.000   -1.36e+04   -4565.203
V2             0.3902      0.009     44.526      0.000       0.373       0.407
V3             0.3186      0.011     30.133      0.000       0.298       0.339
V4            -0.0947      0.018     -5.319      0.000      -0.130      -0.060
V5             0.5268      0.025     20.734      0.000       0.477       0.577
V6            -0.0883      0.023     -3.844      0.000      -0.133      -0.043
V7           484.7038     87.295      5.552      0.000     313.559     655.848
==============================================================================
Omnibus:                     3740.589   Durbin-Watson:                   1.835
Prob(Omnibus):                  0.000   Jarque-Bera (JB):         13015756.569
Skew:                           2.704   Prob(JB):                         0.00
Kurtosis:                     273.547   Cond. No.                     3.30e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.3e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

AIC BIC is lower,the model is better.